### PR TITLE
New functionality that supports compile time branching based on user created dropdowns.

### PR DESCRIPTION
**Pull request by russoh**
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
This PR adds support for branching based on user created dropdowns.

Motivation for work: https://docs.google.com/document/d/1XeUfc2MauivBYx82V68c_u2aQBZJxt2h9VRVu9yemGY/edit

The workflow, and UI for creating dropdowns is very similar to the behavior for working with keyword enums. Dropdowns are a new type of blackboard property, where the user can specify a list of options, and a default. All of the scoping, and keyword specific controls are stripped down, so it is simpler than keywords. The user can drag a dropdown property out onto the graph canvas, which will create a corresponding branch node, with slots for each option.

The key distinction for dropdowns - dropdowns are exposed on the UI of subgraph node instances, where those subgraphs have defined dropdowns. This allows the user to pick the desired branch per subgraph instance from the parent graph. This is all resolved at compile time.

https://jira.unity3d.com/browse/GSG-354

---
### Testing status

Manual testing:

Utilized in Surface Gradient example "Sphere Dropdown And Connection Testing":
https://github.cds.internal.unity3d.com/unity/surfgrad-examples

Automated tests:

Graphics/TestProjects/ShaderGraph/UtilityNodes.unity : Dropdowns
https://jira.unity3d.com/browse/GSG-396

---
### Comments to reviewers

Basic video of this in action:
[SubgraphDropdown.zip](https://github.com/Unity-Technologies/Graphics/files/6282326/SubgraphDropdown.zip)

For testing, use instructions here for getting the compatible build: https://github.cds.internal.unity3d.com/unity/surfgrad-examples/blob/master/README.md

IMPORTANT - you cannot create more than dropdown using the conventional + via the blackboard. If you do this, any secondarily created properties will not be visible, and the graph will break. You MUST make new properties by duplicating the first dropdown using Ctrl-D, until this bug has been addressed. The issue also affects keyword enums, and will be addressed by a separate PR as it is common to both.

I will merge in latest of master periodically to get ready for final merge.

Version numbers of serialized data will need bumping between landing of PRs.

### CODE REVIEW

**Comment by LandonTownsendUnity** [comment]
Test doc: https://docs.google.com/document/d/1-tZ5T0KH_u2ZZE4Fxd8B7bs7QghTkHlj3tAa3e9eSb0/edit?usp=sharing

**Review by joshua-davis — COMMENTED** [review]
One other small note: When you re-order the items in the keywords the Default value seems to point to whatever's at the index, not the item by name. That is if you have:
{ItemA, ItemB, ItemC} and the default is currently ItemB and you then reorder to {ItemA, ItemC, ItemB} then the default will now be ItemC (both index 1). Is this intended? Not a big deal either way, just something I noticed.

File: com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        public override bool allowedInMainGraph { get => false; }
+
+        [SerializeField]
+        private int m_Value;
+
+        public int value
+        {
+            get => m_Value;
+            set => m_Value = value;
```
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs:38]: I noticed the reordering thing, same thing happens for keyword enums, so I just decided to leave it as-as, since it's super easy to fix the default right there when you're changing the ordering. The one thing I **did** specifically address is reordering items affecting the selected dropdown in parent graphs. I don't store by index like enums do, so dependent graphs should not be affected by a simple reordering of the items, like they do for materials.

Regarding m_Value being out of range, I'll have another look today and see if that's a scenario that could occur when indexing into the list, and fix it as needed.

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/DropdownNode.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                sb.AppendLine(string.Format($"{GetVariableNameForSlot(OutputSlotId)} = {value};"));
+            }
+            else
+            {
+                // Iterate all entries in the dropdown
+                for (int i = 0; i < dropdown.entries.Count; i++)
+                {
+                    if (i == 0)
+                    {
+                        sb.AppendLine(string.Format($"{outputSlot.concreteValueType.ToShaderString()} {GetVariableNameForSlot(OutputSlotId)};"));
```
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Nodes/Utility/DropdownNode.cs:135]: It's probably a good idea to have an else block with a default, I'll do this, good suggestion.

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/SubGraphNode.cs [hunk 3]
```diff
@@ -220,6 +226,16 @@ public void GenerateNodeCode(ShaderStringBuilder sb, GenerationMode generationMo
...
             }
 
+            var dropdowns = asset.dropdowns;
+            foreach (var dropdown in dropdowns)
+            {
+                var name = GetDropdownEntryName(dropdown.referenceName);
+                if (dropdown.ContainsEntry(name))
+                    arguments.Add(dropdown.IndexOf(name).ToString());
+                else
+                    arguments.Add(dropdown.value.ToString());
```
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Nodes/Utility/SubGraphNode.cs:236]: This would happen if you had a dropdown selection in a parent graph, and then deleted the entry from the subgraph.

E.g. 
Subgraph dropdown
A, B, C

Parent graph picks B, and serialize.

Now delete B from subgraph, and save asset.

Now parent graph selection now longer exists, so just goes back to default value of dropdown.

File: com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs:38]: Where do you think this might need clamping? I'm not sure I see a failure point for this.

[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs:38]: By the way, I added a fix for reordering thing, so prevent the default value changing.

[the remark this answers is not in the document]
Comment (joshua-davis) [reply com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs:38]:
```
public string entryName
{
    get => entries[value].displayName;
}
```
uses value unguarded so if someone ever sets value to something invalid it will break. I'm not saying this does happen currently, but this is an api that allows things to break.

[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/ShaderDropdown.cs:38]: Alright, I'll put a clamp in there to be sure.

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs [hunk 4]
```diff
@@ -1450,10 +1457,203 @@ string GetSanitizedReferenceName(string name)
...
+            });
+
+            AddPropertyRowToSheet(propertySheet, field, "Default");
+
+            var container = new IMGUIContainer(() => OnDropdownGUIHandler()) { name = "ListContainer" };
+            AddPropertyRowToSheet(propertySheet, container, "Entries");
+            container.SetEnabled(true);
+        }
+
+        void OnDropdownGUIHandler()
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1500]: image

The name of the re-orderable list seems to be uninitialized for some reason 🤔

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs [hunk 5]
```diff
@@ -1450,10 +1457,203 @@ string GetSanitizedReferenceName(string name)
...
+        }
+
+        void DropdownSelectEntry(ReorderableList list)
+        {
+            m_DropdownSelectedIndex = list.index;
+        }
+
+        int GetFirstUnusedDropdownID()
+        {
+            if (!(shaderInput is ShaderDropdown dropdown))
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1589]: This check might be redundant given that you also check for if this input is a dropdown prior to the IMGUI code that calls this and also in the other place you call this function (DropdownAddEntry)

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs [hunk 6]
```diff
@@ -1450,10 +1457,203 @@ string GetSanitizedReferenceName(string name)
...
+
+            // Update GUI
+            this._postChangeValueCallback(true);
+            this._dropdownChangedCallback();
+            m_DropdownSelectedIndex = list.list.Count - 1;
+        }
+
+        void DropdownRemoveEntry(ReorderableList list)
+        {
+            if (!(shaderInput is ShaderDropdown dropdown))
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1628]: Hmm in fact given that you check for whether the input is a dropdown at the entry point to all of this code (i.e. BuildDropdownFields) all these checks after that point might be redundant? unless they are also meant to function as null checks in case a dropdown is somehow invalidated suddenly

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs [hunk 7]
```diff
@@ -721,14 +738,18 @@ public void ToSubGraph()
 
             // Collect the keyword nodes and get the corresponding keywords
             var keywordNodes = graphView.selection.OfType<IShaderNodeView>().Where(x => (x.node is KeywordNode)).Select(x => ((KeywordNode)x.node).keyword);
+            var dropdownNodes = graphView.selection.OfType<IShaderNodeView>().Where(x => (x.node is DropdownNode)).Select(x => ((DropdownNode)x.node).dropdown);
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:741]: Given that dropdowns can only ever be present in a sub graph, could we skip this statement if we're in a parent graph? could save a little bit of execution time on the LINQ expression

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs [hunk 8]
```diff
@@ -721,14 +738,18 @@ public void ToSubGraph()
...
+
             var metaKeywords = graphView.graph.keywords.Where(x => keywordNodes.Contains(x));
+            var metaDropdowns = graphView.graph.dropdowns.Where(x => dropdownNodes.Contains(x));
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:744]: Same comment as above

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs (hunk 7 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:741]: Might be a premature optimization if the branch costs more than the search expression itself, so might be better to profile it before saying one way or the other

**Review by Nightmask3 — CHANGES_REQUESTED** [review]

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 9]
```diff
@@ -2039,6 +2145,21 @@ public override void OnAfterDeserialize(string json)
                         keywordGuidMap[input0.m_Guid.m_GuidSerialized] = keyword;
                     }
 
+                    // Do we even need to worry about this? It is possible that somebody way be using an old version, and later upgrade?
```
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:2149]: Sounds good, will do.

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs (hunk 7 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:741]: I think this is inside the call to "ToSubGraph". How does that get invoked? I could not find a call to it

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs (hunk 6 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1628]: Possibly, I just followed the pattern used by keywords, so I assumed it might be needed?

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs (hunk 5 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1589]: True, unless someone calls it in the wrong context, since there multiple types handled in here. I would prefer to leave this matching the pattern used by keywords to be honest.

Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1589]: Generally the extra checks are probably overly conservative, but it's trivial code, so I would not be concerned about performance implications.

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs (hunk 4 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1500]: Are you referring to the "IList"?

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs [hunk 10]
```diff
@@ -1450,10 +1457,203 @@ string GetSanitizedReferenceName(string name)
...
+        void BuildDropdownField(PropertySheet propertySheet, ShaderDropdown dropdown)
+        {
+            // Clamp value between entry list
+            int value = Mathf.Clamp(dropdown.value, 0, dropdown.entries.Count - 1);
+
+            // Default field
+            var field = new PopupField<string>(dropdown.entries.Select(x => x.displayName).ToList(), value);
+            field.RegisterValueChangedCallback(evt =>
+            {
+                this._preChangeValueCallback("Change Dropdown Value");
```
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1487]: Yep, you're right, that's missing, I'll fix that.

[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1487]: This should work correctly now.

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs (hunk 4 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1500]: I fixed this, added the header, and it now says "Entry Name" instead of "IList".

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs (hunk 6 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1636]: Would prefer to leave this in place out of caution.

File: com.unity.shadergraph/Editor/Drawing/ViewModels/BlackboardViewModel.cs [hunk 11]
```diff
@@ -17,21 +17,25 @@ class BlackboardViewModel : ISGViewModel
         public Dictionary<string, IGraphDataAction> propertyNameToAddActionMap { get; set; }
         public Dictionary<string, IGraphDataAction> defaultKeywordNameToAddActionMap { get; set; }
         public Dictionary<string, IGraphDataAction> builtInKeywordNameToAddActionMap { get; set; }
+        public Dictionary<string, IGraphDataAction> defaultDropdownNameToAddActionMap { get; set; }
```
[the remark this answers is not in the document]
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/ViewModels/BlackboardViewModel.cs]: I changed this to a Tuple.

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs (hunk 7 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:741]: If you select a group of nodes, right click and then in the "Convert To" menu, you'll find the "To SubGraph" option

File: com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs (hunk 4 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs:1500]: Yup that was it

**Review by Nightmask3 — APPROVED** [review]
One last minor thing to possibly look into is the possibility of skipping a LINQ search expression for dropdowns and meta-dropdowns in the `ToSubGraph` function if we're in a non sub-graph, but apart from that looks good, approving!

File: com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs (hunk 7 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Drawing/MaterialGraphEditWindow.cs:741]: Got it, I see now. IMO it makes most sense to leave this generic, and use the logic that already exists to allow/prevent invalid nodes being pasted into the corresponding target graph. I don't think there's value in adding extra complexity to the code here. It's a one time action, and I don't see any performance motivation for doing this.

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/DropdownNode.cs [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+                        sb.AppendLine($"if ({m_Dropdown.value.referenceName} == {i})");
+                    }
+                    else
+                    {
+                        sb.AppendLine($"else if ({m_Dropdown.value.referenceName} == {i})");
+                    }
+
+                    {
+                        sb.AppendLine("{");
```
Comment (joshua-davis) [inline com.unity.shadergraph/Editor/Data/Nodes/Utility/DropdownNode.cs:144]: nit: you could use sb.BlockScope here to clean this up a bit. In particular this becomes:
```
using(sb.BlockScope())
{
    var value = GetSlotValue(GetSlotIdForPermutation(new KeyValuePair<ShaderDropdown, int>(dropdown, i)), generationMode);
    sb.AppendLine(string.Format($"{GetVariableNameForSlot(OutputSlotId)} = {value};"));
}
```
Something I just noticed that I didn't before.

**Review by joshua-davis — APPROVED** [review]
Small nit, but otherwise looks good

**Review by LandonTownsendUnity — APPROVED** [review]
All issues found during my testing were addressed.

**Comment by russoh** [comment]
This one has already had a full green test panel on Shader Graph ABV with the changes specific to this branch:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Fsurfgrad%252Fdropdowns/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/6477815/job

**Comment by russoh** [comment]
ScriptableRenderPipelinePrivate console passes for this PR: https://yamato.cds.internal.unity3d.com/jobs/339-ScriptableRenderPipelinePrivate/tree/sg%252Fsurfgrad%252Fdropdowns/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/6557856/job/pipeline

PR to land changes to ScriptableRenderPipelinePrivate: https://github.cds.internal.unity3d.com/unity/ScriptableRenderPipelinePrivate/pull/168

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 13]
```diff
@@ -1367,16 +1398,23 @@ public string SanitizeGraphInputReferenceName(ShaderInput input, string desiredN
             {
                 case AbstractShaderProperty property:
                 {
-                    // must deduplicate ref names against both keyword and properties, as they occupy the same name space
-                    var existingNames = properties.Where(p => p != property).Select(p => p.referenceName).Union(keywords.Select(p => p.referenceName));
+                    // must deduplicate ref names against keywords, dropdowns, and properties, as they occupy the same name space
+                    var existingNames = properties.Where(p => p != property).Select(p => p.referenceName).Union(keywords.Select(p => p.referenceName)).Union(dropdowns.Select(p => p.referenceName));
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1402]: yikes.  we need a better way to do unique names in the future.  :)

File: com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs [hunk 14]
```diff
@@ -15,6 +15,8 @@ class AbstractMaterialNode0
 
         public string m_KeywordGuidSerialized;
 
+        public string m_DropdownGuidSerialized;
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs:18]: Why does this need to be added to the legacy struct?  This should only be for upgrading from old graphs, which should never have the dropdown...

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 15]
```diff
@@ -2063,6 +2184,11 @@ public override void OnAfterDeserialize(string json)
                             keywordField.SetValue(node, (JsonRef<ShaderKeyword>)keyword);
                         }
 
+                        if (!string.IsNullOrEmpty(node0.m_DropdownGuidSerialized) && dropdownGuidMap.TryGetValue(node0.m_DropdownGuidSerialized, out var dropdown))
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:2187]: shouldn't need the changes in this file -- legacy nodes should never have dropdowns to upgrade?

**Review by cdxntchou — CHANGES_REQUESTED** [review]
Should probably remove the unnecessary versioning changes.   The rest looks good!

File: com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs [hunk 16]
```diff
@@ -21,7 +21,7 @@ class DynamicVectorMaterialSlot : MaterialSlot, IMaterialSlotHasValue<Vector4>
 
         static readonly string[] k_Labels = {"X", "Y", "Z", "W"};
 
-        private ConcreteSlotValueType m_ConcreteValueType = ConcreteSlotValueType.Vector4;
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs:24]: Could you explain why this change was needed?

Comment (elizabeth-legros) [reply com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs:24]: Also, will this give any weird behavior change to our math nodes and the like?

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 17]
```diff
@@ -1238,6 +1244,16 @@ public void CollectShaderKeywords(KeywordCollector collector, GenerationMode gen
             collector.CalculateKeywordPermutations();
         }
 
+        public bool IsInputAllowedInGraph(ShaderInput input)
+        {
+            return (isSubGraph && input.allowedInSubGraph) || (!isSubGraph && input.allowedInMainGraph);
+        }
+
+        public bool IsInputAllowedInGraph(AbstractMaterialNode node)
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1252]: Is this input based or node based? Confused by function signature

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 18]
```diff
@@ -1562,6 +1624,22 @@ public void OnKeywordChangedNoValidate()
...
 
+        public void OnDropdownChanged()
+        {
+            OnDropdownChangedNoValidate();
+            ValidateGraph();
+        }
+
+        public void OnDropdownChangedNoValidate()
+        {
+            var allNodes = GetNodes<AbstractMaterialNode>();
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1635]: Validate graph should do this already, may be redundant

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs [hunk 19]
```diff
@@ -2039,6 +2146,20 @@ public override void OnAfterDeserialize(string json)
                         keywordGuidMap[input0.m_Guid.m_GuidSerialized] = keyword;
                     }
 
+                    foreach (var serializedDropdown in graphData0.m_SerializedDropdowns)
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:2149]: Do we need to handle dropdowns in old master node graphs coming in? Are we backporting dropdowns to master nodes?

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 17 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1252]: Actually there's a function for both - there are guards on properties (for stuff copied from the blackboard between graphs), and for nodes (e.g. the dropdown that is related to the property).

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 18 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1635]: I think in most of those cases, I probably just followed the template used by keywords, but this is possible.

File: com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs (hunk 14 above)
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs:18]: again, will dropdowns be ported to old master nodes? If no, not needed

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 19 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:2149]: This is a good question, should definitely get clarity on that flow.

File: com.unity.shadergraph/Editor/Data/Legacy/GraphData0.cs [hunk 20]
```diff
@@ -17,6 +17,8 @@ class GraphData0
 
         public List<SerializationHelper.JSONSerializedElement> m_SerializedKeywords;
 
+        public List<SerializationHelper.JSONSerializedElement> m_SerializedDropdowns;
```
Comment (elizabeth-legros) [inline com.unity.shadergraph/Editor/Data/Legacy/GraphData0.cs:20]: if not backporting, unnecessary

File: com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs (hunk 16 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/DynamicVectorMaterialSlot.cs:24]: The slot was behaving inconsistently. Initially it would show a 4 channel initializer (0,0,0,0). If you then connected to it (any node), and then disconnected again, the slot would turn into a single channel initializer. With this change it always shows a default single channel initializer, so you don't have that disparity.
Also worth noting that other nodes that accept a dynamic vector, do already use a single channel initializer, so this was notably the odd one out.

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 19 above)
Comment (elizabeth-legros) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:2149]: If we are not (which I would hope at this point tbh), then all the graphdata0 or abstractmaterialnode0 stuff should be unnecessary

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 18 above)
Comment (elizabeth-legros) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1635]: eh, Id say leave it. We need to fix our validation/concretization/etc to be smarter anyways.

**Review by elizabeth-legros — APPROVED** [review]

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 13 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs:1424]: Yes, this is a bit rough! A helper function would simplify this for sure.

File: com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs (hunk 14 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs]: I've removed this one.

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 15 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs]: I've removed this one.

File: com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs (hunk 19 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Graphs/GraphData.cs]: Cheers, I have removed this now.

File: com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs (hunk 14 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Legacy/AbstractMaterialNode0.cs]: Also removed, cheers.

File: com.unity.shadergraph/Editor/Data/Legacy/GraphData0.cs (hunk 20 above)
Comment (russoh) [reply com.unity.shadergraph/Editor/Data/Legacy/GraphData0.cs]: Removed this one.

**Comment by russoh** [comment]
All test have now passed. I will only pull to resolve the Changelog conflict.
